File: PublicFieldRefactoring/PublicFieldRefactoring/CodeRefactoringProvider.cs
using System.Collections.Generic;
using System.Composition;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CodeRefactorings;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Rename;
using Microsoft.CodeAnalysis.Text;

namespace PublicFieldRefactoring
{
    [ExportCodeRefactoringProvider(LanguageNames.CSharp, 
        Name = nameof(PublicFieldRefactoringCodeRefactoringProvider)), Shared]
    internal class PublicFieldRefactoringCodeRefactoringProvider : CodeRefactoringProvider
    {
        public sealed override async Task 
            ComputeRefactoringsAsync(
            CodeRefactoringContext context)
        {
            // Get the root syntax node
            var root = await context.Document.
                GetSyntaxRootAsync(context.CancellationToken).
                ConfigureAwait(false);

            // Find the node at the selection.
            var node = root.FindNode(context.Span);

            // Convert the node into a field declaration
            var fieldDecl = node as FieldDeclarationSyntax;
            // If it's not public, return
            if (fieldDecl == null || 
                fieldDecl.Modifiers.ToFullString().
                Contains("public") == false) { return; }

            // Used to determine if an action must
            // be registered
            bool mustRegisterAction = false;

            // If at least one starting character is
            // lower case, must register an action
            if (fieldDecl.Declaration.
                Variables.Any(i => 
                char.IsLower(i.Identifier.
                ValueText.ToCharArray().First())))
            {
                mustRegisterAction = true;
            }
            else
            {
                mustRegisterAction = false;
            }

            
            if (mustRegisterAction==false)
            {
                return;
            }

            // Create an action invoking a delegate passing
            // the document instance, the syntax node, and
            // a cancellation token
            var action = 
                CodeAction.Create("Make first char upper case",
                           c => RenameFieldAsync(context.
                           Document, fieldDecl, c));

            // Register this code action
            context.RegisterRefactoring(action);       
        }

        private async Task<Document> 
            RenameFieldAsync(Document document, 
            FieldDeclarationSyntax fieldDeclaration, 
            CancellationToken cancellationToken)
        {
            // Get the semantic model for the code file
            var semanticModel = 
                await document.
                      GetSemanticModelAsync(cancellationToken).
                      ConfigureAwait(false);

            // Get the root syntax node
            var root = await document.GetSyntaxRootAsync();

            // Get a list of old variable declarations
            var oldDeclarators = 
                fieldDeclaration.Declaration.
                Variables;

            // Store the collection of variables
            var listOfNewModifiedDeclarators = 
                new SeparatedSyntaxList<
                    VariableDeclaratorSyntax>();

            // Iterate the declarators collection
            foreach (var declarator in oldDeclarators) {
                // Get a new name
                var tempString = 
                    ConvertName(declarator.Identifier.
                    ToFullString());

                // Generate a new modified declarator
                // based on the previous one but with
                // a new identifier
                listOfNewModifiedDeclarators = 
                    listOfNewModifiedDeclarators.
                    Add(declarator.WithIdentifier(
                        SyntaxFactory.ParseToken(tempString)));
            }

            // Generate a new field declaration
            // with updated variable names
            var newDeclaration = 
                fieldDeclaration.Declaration.
                WithVariables(listOfNewModifiedDeclarators);

            // Generate a new FieldDeclarationSyntax
            var newField = fieldDeclaration.
                WithDeclaration(newDeclaration);

            // Replace the old syntax node with the new one
            SyntaxNode newRoot = root.
                       ReplaceNode(fieldDeclaration, 
                       newField);

            // Generate a new document
            var newDocument = 
                document.WithSyntaxRoot(newRoot);

            // Return the document
            return newDocument;
       }

        // Return a new identifier with an
        // uppercase letter
        private string ConvertName(string oldName)
        {
            return char.
                   ToUpperInvariant(oldName[0]) + 
                   oldName.Substring(1);
        }
    }
}

File: PublicFieldRefactoring_VB/PublicFieldRefactoring_VB/CodeRefactoringProvider.vb
<ExportCodeRefactoringProvider(LanguageNames.VisualBasic, Name:=NameOf(PublicFieldRefactoring_VBCodeRefactoringProvider)), [Shared]>
Friend Class PublicFieldRefactoring_VBCodeRefactoringProvider
    Inherits CodeRefactoringProvider

    Public NotOverridable Overrides Async _
        Function ComputeRefactoringsAsync(
                 context As CodeRefactoringContext) As Task

        'Get the root syntax node
        Dim root = Await context.Document.
            GetSyntaxRootAsync(context.
            CancellationToken).ConfigureAwait(False)

        ' Find the node at the selection.
        Dim node = root.FindNode(context.Span)

        'Convert the node to a field declaration
        Dim fieldDecl = TryCast(node,
            FieldDeclarationSyntax)

        'If it's not Public, return
        If fieldDecl Is Nothing Or fieldDecl.Modifiers.
            ToFullString.Contains("Public") = False Then
            Return
        End If

        'Used to determine if an action
        'must be registered
        Dim mustRegisterAction As Boolean

        'Iterate the variable declarators
        For Each declarator
            In fieldDecl.Declarators

            'If at least one starting character 
            'is lower case, must register an action 
            If declarator.Names.Any(Function(d) _
                          Char.IsLower(d.Identifier.
                          Value.ToString(0))) Then
                mustRegisterAction = True
            Else
                mustRegisterAction = False
            End If
        Next

        If mustRegisterAction = False Then
            Return
        Else
            'Create an action invoking a delegate passing
            'the document instance, the syntax node, and
            'a cancellation token
            Dim action =
                CodeAction.Create("Make first char upper case",
                           Function(c) RenameFieldAsync(context.
                           Document, fieldDecl, c))

            ' Register this code action.
            context.RegisterRefactoring(action)
        End If
    End Function

    Private Async Function _
        RenameFieldAsync(document As Document,
                         fieldDeclaration As FieldDeclarationSyntax,
                         cancellationToken As CancellationToken) _
                         As Task(Of Document)

        'Get the semantic model for the code file
        Dim semanticModel = Await document.
            GetSemanticModelAsync(cancellationToken).
            ConfigureAwait(False)

        'Get the root syntax node
        Dim root = Await document.GetSyntaxRootAsync

        'Get a list of old declarators
        Dim oldDeclarators =
            fieldDeclaration.Declarators

        'Store the collection of identifiers
        Dim listOfNewModifiedIdentifiers As _
            New SeparatedSyntaxList(
            Of ModifiedIdentifierSyntax)
        'Store the collection of declarators
        Dim listOfNewModifiedDeclarators As _
            New SeparatedSyntaxList(
            Of VariableDeclaratorSyntax)

        'Iterate the declarators collection
        For Each declarator In oldDeclarators
            'For each variable name in the declarator...
            For Each modifiedIdentifier In declarator.Names
                'Get a new name
                Dim tempString =
                    ConvertName(modifiedIdentifier.
                    ToFullString())

                'Generate a new ModifiedIdentifierSyntax based on
                'the previous one's properties but with a new Identifier
                Dim newModifiedIdentifier As _
                    ModifiedIdentifierSyntax =
                    modifiedIdentifier.
                    WithIdentifier(SyntaxFactory.
                    ParseToken(tempString)).
                    WithTrailingTrivia(modifiedIdentifier.
                    GetTrailingTrivia)

                'Add the new element to the collection
                listOfNewModifiedIdentifiers =
                    listOfNewModifiedIdentifiers.
                    Add(newModifiedIdentifier)
            Next
            'Store a new variable declarator with new
            'variable names
            listOfNewModifiedDeclarators =
                listOfNewModifiedDeclarators.Add(declarator.
                WithNames(listOfNewModifiedIdentifiers))

            'Clear the list before next iteration
            listOfNewModifiedIdentifiers = Nothing
            listOfNewModifiedIdentifiers = New _
                SeparatedSyntaxList(Of ModifiedIdentifierSyntax)
        Next

        'Generate a new FieldDeclarationSyntax
        Dim newField = fieldDeclaration.
            WithDeclarators(listOfNewModifiedDeclarators)

        'Replace the old declaration with the new one
        Dim newRoot As SyntaxNode =
            root.ReplaceNode(fieldDeclaration,
                             newField)

        'Generate a new document
        Dim newDocument =
            document.WithSyntaxRoot(newRoot)

        'Return the new document
        Return newDocument
    End Function

    'Return a new identifier with an
    'uppercase letter
    Private Function _
        ConvertName(oldName As String) As String
        Return Char.
            ToUpperInvariant(oldName(0)) &
            oldName.Substring(1)
    End Function
End Class